### PR TITLE
expression: refine builtin-func RAND (fix issue3211)

### LINKED ISSUE #3211: Wrong select random() statement result.
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
I hope to get random Date from select statement.
select DATE_ADD("2016-01-01", INTERVAL RAND(20160101) * 364 DAY) from (select 1 union all select 2 union all select 3 union all select 4) a;

2. What did you expect to see?
MariaDB Result:
+-----------------------------------------------------------+
| DATE_ADD("2016-01-01", INTERVAL RAND(20160101) * 364 DAY) |
+-----------------------------------------------------------+
| 2016-11-25                                                |
| 2016-11-20                                                |
| 2016-09-23                                                |
| 2016-12-24                                                |
+-----------------------------------------------------------+
4 rows in set (0.00 sec)
Differrent date.

3. What did you see instead?
TiDB result:
+-----------------------------------------------------------+
| DATE_ADD("2016-01-01", INTERVAL RAND(20160101) * 364 DAY) |
+-----------------------------------------------------------+
| 2016-12-14                                                |
| 2016-12-14                                                |
| 2016-12-14                                                |
| 2016-12-14                                                |
+-----------------------------------------------------------+
4 rows in set (0.00 sec)
4 same Date.

4. What version of TiDB are you using (`tidb-server -V`)?
2017/05/05 01:33:38 printer.go:31: [info] Welcome to TiDB.
2017/05/05 01:33:38 printer.go:32: [info] Version:
2017/05/05 01:33:38 printer.go:33: [info] Git Commit Hash: ec4eb933fa6786473ee31485c3d317ede63da5ac
2017/05/05 01:33:38 printer.go:34: [info] UTC Build Time:  2017-05-02 03:02:10

I don't know which policy for ConstantFold, but I suggest do more test case.

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit fixes #3211 through recording the state of RAND built-in function.

``` sql
create table t (a int);
insert into t values(1),(2),(3),(4);
select rand(20160101) from t;
```

Before this commit, the query above gets 4 same results,
because rand(20160101) executes 
``` go
rand.Seed(seed)
rand.Float64() 
```
repeatedly.

@shenli @coocood PTAL

### CODE REVIEW

File: expression/builtin_math.go [hunk 1]
```diff
@@ -341,12 +346,20 @@ func (b *builtinRandSig) eval(row []types.Datum) (d types.Datum, err error) {
...
 		if err != nil {
 			return d, errors.Trace(err)
 		}
 		rand.Seed(seed)
+		cnt := args[1].GetInt64()
+		// To avoid get a same result every time for `rand(seed)`,
+		// we check arg[1] to get the time that RAND has executed,
+		// and skip the first cnt results.
+		for i := 0; i < int(cnt); i++ {
+			rand.Float64()
```
Comment (shenli) [inline expression/builtin_math.go]: Can we save rand in builtinRandSig as a member? So we could save the loop.

**Comment by coocood** [comment]
@XuHuaiyu 
I think a better solution is to create a rand instance by `rand.New()` and use it for the particalur function.

**Comment by coocood** [comment]
If the seed is not given, we use the current timestamp as the seed, no need to append an argument.

**Comment by XuHuaiyu** [comment]
PTAL @coocood @shenli

File: expression/builtin_math.go [hunk 2]
```diff
@@ -342,13 +343,17 @@ func (b *builtinRandSig) eval(row []types.Datum) (d types.Datum, err error) {
...
+			seed, err := args[0].ToInt64(b.ctx.GetSessionVars().StmtCtx)
+			if err != nil {
+				return d, errors.Trace(err)
+			}
+			b.ranGen = rand.New(rand.NewSource(seed))
 		}
-		rand.Seed(seed)
+		d.SetFloat64(b.ranGen.Float64())
+	} else {
+		d.SetFloat64(rand.Float64())
```
Comment (coocood) [inline expression/builtin_math.go]: We should always use `randGen`.

File: expression/builtin_math.go [hunk 3]
```diff
@@ -325,13 +325,14 @@ type randFunctionClass struct {
...
 	err := errors.Trace(c.verifyArgs(args))
-	bt := &builtinRandSig{newBaseBuiltinFunc(args, ctx)}
+	bt := &builtinRandSig{baseBuiltinFunc: newBaseBuiltinFunc(args, ctx)}
 	bt.deterministic = false
 	return bt, errors.Trace(err)
 }
 
 type builtinRandSig struct {
 	baseBuiltinFunc
+	ranGen *rand.Rand
```
Comment (coocood) [inline expression/builtin_math.go]: s/ranGen/randGen

`ran` may represent `range`.

File: expression/builtin_math_test.go [hunk 4]
```diff
@@ -188,6 +190,16 @@ func (s *testEvaluatorSuite) TestRand(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(v.GetFloat64(), Less, float64(1))
 	c.Assert(v.GetFloat64(), GreaterEqual, float64(0))
+
+	// issue 3211
+	f2, err := fc.getFunction([]Expression{&Constant{Value: types.NewIntDatum(20160101), RetType: types.NewFieldType(mysql.TypeLonglong)}}, s.ctx)
+	c.Assert(err, IsNil)
+	rand.Seed(20160101)
```
Comment (coocood) [inline expression/builtin_math_test.go]: Do not change the global random instance.

**Comment by shenli** [comment]
Much better!
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

File: expression/builtin_math.go [hunk 5]
```diff
@@ -29,6 +29,7 @@ import (
 	"github.com/pingcap/tidb/context"
 	"github.com/pingcap/tidb/parser/opcode"
 	"github.com/pingcap/tidb/util/types"
+	"time"
```
Comment (zimulala) [inline expression/builtin_math.go]: Move this line to line 27.

File: expression/builtin_math_test.go [hunk 6]
```diff
@@ -18,9 +18,11 @@ import (
 
 	. "github.com/pingcap/check"
 	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/util/testleak"
 	"github.com/pingcap/tidb/util/testutil"
 	"github.com/pingcap/tidb/util/types"
+	"math/rand"
```
Comment (zimulala) [inline expression/builtin_math_test.go]: Ditto.

**Pull request merged**